sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetValidation.java

```diff
@@ -0,0 +1,292 @@
+        }
+      } else if (c2 == 0) {
+        leAscii++;
+        if (c1 == 0x0a || c1 == 0x0d) {
+          leLines++;
+        }
+      }
+    }
+
+    float beAsciiPerc = beAscii * 2.0f / (float) buffer.length;
+    float leAsciiPerc = leAscii * 2.0f / (float) buffer.length;
+
+    if (leLines == 0) {
+      // could be BE
+      if (beAsciiPerc >= UTF_16_NULL_PASS_THRESHOLD && leAsciiPerc < UTF_16_NULL_FAIL_THRESHOLD) {
+        return Result.newValid(StandardCharsets.UTF_16BE);
+      }
+      if (beLines > 0) {
+        // this gives FPs for UTF-32 if !failOnNull
+        return Result.newValid(StandardCharsets.UTF_16BE);
+      }
+    } else if (beLines > 0) {
+      // lines detected with both endiness -> can't be utf-16
+      return Result.newInvalid();
+    }
+    if (beLines == 0) {
+      // could be BE
+      if (leAsciiPerc >= UTF_16_NULL_PASS_THRESHOLD && beAsciiPerc < UTF_16_NULL_FAIL_THRESHOLD) {
+        return Result.newValid(StandardCharsets.UTF_16LE);
+      }
+      if (leLines > 0) {
+        // this gives FPs for UTF-32 if !failOnNull
+        return Result.newValid(StandardCharsets.UTF_16LE);
+      }
+    }
+
+    // if we reach here, means that there wasn't a line feed for a single endiness and we didn't see a strong null pattern for any of the
+    // endiness.
+    // It could happen if there are no line feeds in the text and it's a language that does not use ANSI (unicode > 255).
+    return new Result(Valid.MAYBE, null);
+  }
+
+  public boolean looksBinary() {
+    return false;
+  }
+
+  /**
+   * Checks whether it's a valid UTF-16-encoded buffer. 
+   * Most sequences of bytes of any encoding will be valid UTF-16, so this is not very effective and gives
+   * often false positives.
+   * 
+   * Possible 16bit values in UTF-16:
+   * 
+   * 0x0000-0xD7FF: single 16bit block
+   * 0xD800-0xDBFF: first block
+   * 0xDC00-0xDFFF: second block
+   * 0XE000-0xFFFF: single 16 bit block
+   * 
+   * The following UTF code points get mapped into 1 or 2 blocks:
+   * 0x0000 -0xD7FF   (0    -55295)  : 2 bytes, direct mapping
+   * 0xE000 -0xFFFF   (57344-65535)  : 2 bytes, direct mapping
+   * 0x10000-0x10FFFF (65536-1114111): 2 blocks of 2 bytes (not direct..)
+   * 
+   * Note that Unicode 55296-57345 (0xD800 to 0xDFFF) are not used, since it's reserved and used in UTF-16 for the high/low surrogates.
+   * 
+   * We reject 2-byte blocks with 0 (we consider it's binary) even though it's a valid UTF-16 encoding.
+   * 
+   */
+  public boolean isValidUTF16(byte[] buffer) {
+    return isValidUTF16(buffer, false);
+  }
+
+  public boolean isValidUTF16(byte[] buffer, boolean le) {
+    if (buffer.length < 2) {
+      return false;
+    }
+    for (int i = 0; i < buffer.length / 2; i++) {
+      boolean extraByte = false;
+      int c = read16bit(buffer, i, le);
+
+      if (c >= 0xD800 && c < 0xDC00) {
+        // it's a higher surrogate (10 bits)
+        extraByte = true;
+        i++;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor the code in order to not assign to this loop counter from within the loop body. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)